subsys/debug/tracing/cpu_stats.c

```diff
@@ -123,7 +121,6 @@ void sys_trace_thread_switched_out(void)
 	int key = irq_lock();
 
 	__ASSERT_NO_MSG(nested_interrupts == 0);
-	__ASSERT_NO_MSG(current_thread == k_current_get());
```

Comment:
How about, then, adding the dual assert in sys_trace_thread_switched_in?

Reply:
That would reverse the required order of the calls, but as far as I can determine, ARM for one,  is not calling out before in and thus it would then break for ARM. 
I do not see any need for imposing the calling order on these functions, and as it is now, it does not work for initial switch to main for architectures not setting CONFIG_ARCH_HAS_CUSTOM_SWAP_TO_MAIN.